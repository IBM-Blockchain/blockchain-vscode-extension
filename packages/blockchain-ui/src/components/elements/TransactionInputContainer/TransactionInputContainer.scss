.vscode-light {
    .transaction-input-content-switch.bx--content-switcher {
        display: flex;
        justify-content: space-evenly;
        width: 100%;
        height: 2.5rem;

        .bx--content-switcher-btn {
            box-sizing: border-box;
            font-size: 100%;
            font-family: inherit;
            vertical-align: initial;
            font-size: .875rem;
            font-weight: 400;
            line-height: 1.125rem;
            letter-spacing: .16px;
            outline: 2px solid transparent;
            outline-offset: -2px;
            background-color: #f4f4f4;
            display: inline-flex;
            align-items: center;
            width: 100%;
            padding: .5rem 1rem;
            margin: 0;
            white-space: nowrap;
            overflow: hidden;
            text-align: left;
            text-decoration: none;
            border: none;
            color: #161616;
            transition: all 70ms cubic-bezier(.2,0,.38,.9);
            position: relative;
            cursor: pointer;
        }

        .bx--content-switcher-btn *, .bx--content-switcher-btn :after, .bx--content-switcher-btn :before {
            box-sizing: inherit;
        }

        .bx--content-switcher-btn:first-child {
            border-top-left-radius: .25rem;
            border-bottom-left-radius: .25rem;
        }

        .bx--content-switcher-btn.bx--content-switcher--selected {
            background-color: #393939;
            color: #c6c6c6;
            z-index: 3;
        }

        .bx--content-switcher__label {
            white-space: nowrap;
            overflow: hidden;
            text-overflow: ellipsis;
            max-width: 100%;
        }
    }
}