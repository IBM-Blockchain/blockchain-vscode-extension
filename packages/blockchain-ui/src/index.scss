@import '@carbon/colors/scss/colors';
@import '@carbon/themes/scss/themes';
@import '@carbon/type/scss/type';
@import '@carbon/type/scss/font-face/_mono';
@import '@carbon/type/scss/font-face/_sans';
@import './mixins/theme-mixins.scss';

@include carbon--type-reset();
@include carbon--font-face-mono();
@include carbon--font-face-sans();

html {

    .vscode-dark {
        @include carbon--theme($carbon--theme--g90) {

            #root {
                // Color will default to #000000 otherwise
                color: $text-01;        
                @include select-dark;
            }

            // text inputs on the create transaction screen
            #create-txn-form {
                @include text-inputs-dark;
            }

            // data tables on home page
            #recent-txns-table, #saved-txns-table {
                @include data-table-dark;
            }

            // InlineSelect components
            #smart-contract-select-wrapper {
                @include inline-select-dark;
            }

            // Checkbox
            @include checkbox-dark;

            //panel containing transaction output
            #output-panel {
                @include output-panel-dark;
            }
        }
    }

    .vscode-light {
        @include carbon--theme($carbon--theme--g10) {
            
            #root {
                color: $text-01;
            }
        }
    }
}
