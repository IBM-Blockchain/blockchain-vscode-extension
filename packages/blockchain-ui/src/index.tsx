import * as React from 'react';
import * as ReactDOM from 'react-dom';
import App from './App';
import './index.scss';

const app: any = ReactDOM.render(
    <App/>,
    document.getElementById('root')
);

// @ts-ignore
if (window.Cypress) {
    // @ts-ignore
    window.app = app;
}
