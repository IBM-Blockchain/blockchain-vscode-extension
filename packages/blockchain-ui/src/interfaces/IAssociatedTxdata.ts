interface IAssociatedTxdata {
    chaincodeName: string;
    channelName: string;
    transactionDataPath: string;
}

export default IAssociatedTxdata;
