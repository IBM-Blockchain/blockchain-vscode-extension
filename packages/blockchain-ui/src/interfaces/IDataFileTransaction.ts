export interface IDataFileTransaction {
    transactionName: string;
    transactionLabel: string;
    txDataFile: string;
    arguments: Array<string>;
    transientData: object;
}

export default IDataFileTransaction;
