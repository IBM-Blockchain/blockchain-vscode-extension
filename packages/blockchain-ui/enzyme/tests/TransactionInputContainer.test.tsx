// tslint:disable no-unused-expression
import React from 'react';
import {act} from 'react-dom/test-utils';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { ContentSwitcher, Dropdown, MultiSelect, FileUploaderItem } from 'carbon-components-react';
import TransactionInputContainer from '../../src/components/elements/TransactionInputContainer/TransactionInputContainer';
import TransactionManualInput from '../../src/components/elements/TransactionManualInput/TransactionManualInput';
import chai from 'chai';
import sinon from 'sinon';
import sinonChai from 'sinon-chai';
import ITransaction from '../../src/interfaces/ITransaction';
import ISmartContract from '../../src/interfaces/ISmartContract';
import Utils from '../../src/Utils';
import { ExtensionCommands } from '../../src/ExtensionCommands';
import IAssociatedTxdata from '../../src/interfaces/IAssociatedTxdata';
import IDataFileTransaction from '../../src/interfaces/IDataFileTransaction';
import ITransactionManualInput from '../../src/interfaces/ITransactionManualInput';
chai.should();
chai.use(sinonChai);

const transactionNameSelector: any = Dropdown;
const transactionParametersSelector: any = '#arguments-text-area';
const transientDataSelector: any = '#transient-data-input';

function toggleContentSwitcher(component: any): any {
    act(() => {
        component.find(ContentSwitcher).prop('onChange')();
    });
    component = component.update();
    return component;
}

function updateManualInputValues(component: any, transactionName: string | undefined, transactionParameters: string | undefined, transientData: string | undefined): any {
    if (transactionName !== undefined) {
        const transactionInput: any = component.find(transactionNameSelector).at(0);
        act(() => {
            transactionInput.prop('onChange')({ selectedItem: transactionName });
        });
        component = component.update();
    }

    if (transactionParameters !== undefined) {
        const parameterInput: any = component.find(transactionParametersSelector).at(0);
        act(() => {
            parameterInput.prop('onChange')({ currentTarget: { value: transactionParameters }});
        });
        component = component.update();
    }

    if (transientData !== undefined) {
        const transientDataInput: any = component.find(transientDataSelector).at(0);
        act(() => {
            transientDataInput.prop('onChange')({ currentTarget: { value: transientData }});
        });
        component = component.update();
    }

    return component;
}

function updateDataInputValues(component: any, transactions: IDataFileTransaction | undefined): any {
    if (transactions !== undefined) {
        const transactionInput: any = component.find(transactionNameSelector);
        act(() => {
            transactionInput.prop('onChange')({ selectedItem: transactions });
        });
        component = component.update();
    }

    return component;
}

function updateSharedInputValues(component: any, selectedPeers: { id: string; label: string}[]): any {
    if (selectedPeers.length > 0) {

        const peersInput: any = component.find(MultiSelect);
        act(() => {
            peersInput.prop('onChange')({ selectedItems: selectedPeers });
        });

        component = component.update();
    }

    return component;
}

describe('TransactionInputContainer component', () => {
    let mySandbox: sinon.SinonSandbox;
    let postToVSCodeStub: sinon.SinonStub;
    let component: any;

    const transactionOne: ITransaction = {
        name: 'transactionOne',
        parameters: [{
            description: '',
            name: 'name',
            schema: {}
        }],
        returns: {
            type: ''
        },
        tag: ['submit']
    };

    const transactionTwo: ITransaction = {
        name: 'transactionTwo',
        parameters: [],
        returns: {
            type: ''
        },
        tag: ['submit']
    };

    const greenContract: ISmartContract = {
        name: 'greenContract',
        version: '0.0.1',
        channel: 'mychannel',
        label: 'greenContract@0.0.1',
        transactions: [transactionOne, transactionTwo],
        namespace: 'GreenContract',
        peerNames: ['peer1', 'peer2']
    };

    const associatedTxdata: IAssociatedTxdata = {
        chaincodeName: 'chaincodeName',
        channelName: 'channelName',
        transactionDataPath: 'transactionDataPath',
    };

    const txdataTransactions: IDataFileTransaction[] = [{
        transactionName: transactionOne.name,
        transactionLabel: transactionOne.name,
        txDataFile: 'transactionData.txdata',
        arguments: ['arg1', 'arg2'],
        transientData: { key: 'value' }
    }];

    const preselectedTransaction: ITransaction = {
        name: '',
        parameters: [],
        returns: { type: '' },
        tag: [],
    };

    beforeEach(async () => {
        mySandbox = sinon.createSandbox();
        postToVSCodeStub = mySandbox.stub(Utils, 'postToVSCode');
    });

    afterEach(async () => {
        mySandbox.restore();
    });

    it('should render the expected snapshot', () => {
        const snapshotComponent: any = renderer
            .create(<TransactionInputContainer smartContract={greenContract} associatedTxdata={associatedTxdata} txdataTransactions={txdataTransactions} preselectedTransaction={preselectedTransaction} />)
            .toJSON();
        expect(snapshotComponent).toMatchSnapshot();
    });

    describe('Content switcher', () => {
        beforeEach(() => {
            component = mount(<TransactionInputContainer smartContract={greenContract} associatedTxdata={associatedTxdata} txdataTransactions={txdataTransactions} preselectedTransaction={preselectedTransaction} />);
        });

        it('shows the manual input by default', () => {
            const parameterTextArea: any = component.find(transactionParametersSelector);
            const filePicker: any = component.find(FileUploaderItem);
            expect(parameterTextArea.exists()).toBeTruthy();
            expect(filePicker.exists()).toBeFalsy();
        });

        it('shows the data file input when toggled', () => {
            component = toggleContentSwitcher(component);
            const parameterTextArea: any = component.find(transactionParametersSelector);
            const filePicker: any = component.find(FileUploaderItem);
            expect(parameterTextArea.exists()).toBeFalsy();
            expect(filePicker.exists()).toBeTruthy();
        });

        ['content-switch-manual', 'content-switch-data'].map((dataId) => {
            it('switch component does nothing when either the onClick or onKeyDown functions are called', () => {
                let switchComponent: any = component.find(`[data-testid="${dataId}"]`).at(1);
                switchComponent.simulate('click');
                component = component.update();
                switchComponent = component.find(`[data-testid="${dataId}"]`).at(1);
                switchComponent.prop('onKeyDown')({ key: 'Enter' });
                component = component.update();
                expect(component.matchesElement(<TransactionInputContainer smartContract={greenContract} associatedTxdata={associatedTxdata} txdataTransactions={txdataTransactions} preselectedTransaction={preselectedTransaction} />)).toEqual(true);
            });
        });
    });

    describe('Manual input', () => {
        beforeEach(() => {
            component = mount(<TransactionInputContainer smartContract={greenContract} associatedTxdata={associatedTxdata} txdataTransactions={txdataTransactions} preselectedTransaction={preselectedTransaction}/>);
        });

        it('updates the dropdown when the transaction is chosen', () => {
            let dropdown: any = component.find(transactionNameSelector);
            let manualInput: any = component.find(TransactionManualInput);
            let manualInputState: ITransactionManualInput = manualInput.prop('manualInputState');

            expect(dropdown.prop('selectedItem')).toEqual('Select the transaction name');
            expect(manualInputState.activeTransaction).toEqual({ name: '', parameters: [], returns: { type: '' }, tag: [] });

            component = updateManualInputValues(component, 'transactionOne', undefined, undefined);

            dropdown = component.find(transactionNameSelector);
            manualInput = component.find(TransactionManualInput);
            manualInputState = manualInput.prop('manualInputState');
            expect(manualInputState.activeTransaction).toHaveProperty('name', 'transactionOne');
            expect(dropdown.prop('selectedItem')).toEqual('transactionOne');
        });

        it('sets the activeTransaction to the preselectedTransaction prop when it is passed in', () => {
            const selectedTransaction: ITransaction = greenContract.transactions[0];
            // Use a different component as we need to pass in the preselectedTransaction
            component = mount(<TransactionInputContainer smartContract={greenContract} associatedTxdata={associatedTxdata} txdataTransactions={txdataTransactions} preselectedTransaction={selectedTransaction} />);
            const dropdown: any = component.find(transactionNameSelector);
            const manualInput: any = component.find(TransactionManualInput);
            const manualInputState: ITransactionManualInput = manualInput.prop('manualInputState');

            expect(manualInputState.activeTransaction).toEqual(selectedTransaction);
            expect(dropdown.prop('selectedItem')).toEqual(selectedTransaction.name);
        });

        it('generates transaction arguments when an option from the transaction select is chosen', () => {
            component = updateManualInputValues(component, 'transactionOne', undefined, undefined);
            const textarea: any = component.find(transactionParametersSelector).at(0);
            expect(textarea.prop('value')).toEqual('{\n  "name": ""\n}');
            expect(textarea.prop('disabled')).toBeFalsy();
        });

        it('does not generate arguments in the event that the chosen transaction doesn\'t have any parameters', () => {
            component = updateManualInputValues(component, 'transactionTwo', undefined, undefined);
            const textarea: any = component.find(transactionParametersSelector).at(0);
            expect(textarea.prop('value')).toEqual('{}');
            expect(textarea.prop('disabled')).toBeFalsy();
        });

        it('does not generate arguments in the event that the chosen transaction doesn\'t exist', () => {
            component = updateManualInputValues(component, 'anotherTransaction', undefined, undefined);
            const textarea: any = component.find(transactionParametersSelector).at(0);
            expect(textarea.prop('value')).toEqual('{}');
            expect(textarea.prop('disabled')).toBeTruthy();
        });

        it('updates when the user types in the textarea', () => {
            let manualInput: any = component.find(TransactionManualInput);
            let manualInputState: ITransactionManualInput = manualInput.prop('manualInputState');

            expect(manualInputState.transactionArguments).toEqual([]);

            component = updateManualInputValues(component, undefined, '{"key": "the value"}', undefined);

            manualInput = component.find(TransactionManualInput);
            manualInputState = manualInput.prop('manualInputState');
            expect(manualInputState.transactionArguments).toEqual(['the value']);
        });

        it('updates when the user removes all args from the textarea', () => {
            let manualInput: any = component.find(TransactionManualInput);
            let manualInputState: ITransactionManualInput = manualInput.prop('manualInputState');

            expect(manualInputState.transactionArguments).toEqual([]);
            component = updateManualInputValues(component, undefined, '{"key": "the value"}', undefined);

            manualInput = component.find(TransactionManualInput);
            manualInputState = manualInput.prop('manualInputState');
            expect(manualInputState.transactionArguments).toEqual(['the value']);
            component = updateManualInputValues(component, undefined, '', undefined);

            manualInput = component.find(TransactionManualInput);
            manualInputState = manualInput.prop('manualInputState');
            expect(manualInputState.transactionArguments).toEqual([]);
        });

        it('updates when the user types in the transient data input box', () => {
            let manualInput: any = component.find(TransactionManualInput);
            let manualInputState: ITransactionManualInput = manualInput.prop('manualInputState');

            expect(manualInputState.transientData).toEqual('');

            component = updateManualInputValues(component, undefined, undefined, 'some transient data');

            manualInput = component.find(TransactionManualInput);
            manualInputState = manualInput.prop('manualInputState');
            expect(manualInputState.transientData).toEqual('some transient data');
        });

        it('should update state when a peer is selected', () => {
            let multiSelect: any = component.find(MultiSelect);
            let selectedValues: ITransactionManualInput = multiSelect.prop('initialSelectedItems');

            expect(selectedValues).toEqual([{ id: 'peer1', label: 'peer1' }, { id: 'peer2', label: 'peer2' }]);

            component = updateSharedInputValues(component, [{
                id: 'peer1',
                label: 'peer1'
            }]);

            multiSelect = component.find(MultiSelect);
            selectedValues = multiSelect.prop('initialSelectedItems');
            expect(selectedValues).toEqual([{ id: 'peer1', label: 'peer1' }]);
        });

        it('should attempt to submit a transaction when the submit button is clicked ', () => {
            component = updateManualInputValues(component, 'transactionOne', '{"key": "Green"}', undefined);

            component.find('#submit-button').at(0).simulate('click');

            postToVSCodeStub.should.have.been.calledOnceWithExactly({
                command: ExtensionCommands.SUBMIT_TRANSACTION,
                data: {
                    args: ['Green'],
                    channelName: 'mychannel',
                    evaluate: false,
                    namespace: 'GreenContract',
                    peerTargetNames: greenContract.peerNames,
                    smartContract: 'greenContract',
                    transactionName: 'transactionOne',
                    transientData: '',
                    txDataFile: undefined,
                }
            });
        });

        it('should attempt to evaluate a transaction when the evaluate button is clicked', () => {
            component = updateManualInputValues(component, 'transactionOne', '{"key": "Green"}', undefined);

            component.find('#evaluate-button').at(1).simulate('click');
            postToVSCodeStub.should.have.been.calledOnceWithExactly({
                command: ExtensionCommands.EVALUATE_TRANSACTION,
                data: {
                    args: ['Green'],
                    channelName: 'mychannel',
                    evaluate: true,
                    namespace: 'GreenContract',
                    peerTargetNames: greenContract.peerNames,
                    smartContract: 'greenContract',
                    transactionName: 'transactionOne',
                    transientData: '',
                    txDataFile: undefined,
                }
            });
        });

        it('should attempt to submit a transaction with transient data when the submit button is clicked ', () => {
            component = updateManualInputValues(component, 'transactionOne', '{"key": "Green"}', '{"some": "data"}');

            component.find('#submit-button').at(1).simulate('click');
            postToVSCodeStub.should.have.been.calledOnceWithExactly({
                command: ExtensionCommands.SUBMIT_TRANSACTION,
                data: {
                    args: ['Green'],
                    channelName: 'mychannel',
                    evaluate: false,
                    namespace: 'GreenContract',
                    peerTargetNames: greenContract.peerNames,
                    smartContract: 'greenContract',
                    transactionName: 'transactionOne',
                    transientData: '{"some": "data"}',
                    txDataFile: undefined,
                }
            });
        });

        it('should attempt to evaluate a transaction with transient data when the evaluate button is clicked', () => {
            component = updateManualInputValues(component, 'transactionOne', '{"key": "Green"}', '{"some": "data"}');

            component.find('#evaluate-button').at(1).simulate('click');
            postToVSCodeStub.should.have.been.calledOnceWithExactly({
                command: ExtensionCommands.EVALUATE_TRANSACTION,
                data: {
                    args: ['Green'],
                    channelName: 'mychannel',
                    evaluate: true,
                    namespace: 'GreenContract',
                    peerTargetNames: greenContract.peerNames,
                    smartContract: 'greenContract',
                    transactionName: 'transactionOne',
                    transientData: '{"some": "data"}',
                    txDataFile: undefined,
                }
            });
        });

        it('should attempt to submit a transaction with custom peers when the submit button is clicked', () => {
            component = updateManualInputValues(component, 'transactionOne', '{"key": "Green"}', '{"some": "data"}');

            component = updateSharedInputValues(component, [{
                id: 'peer1',
                label: 'peer1'
            }]);

            component.find('#submit-button').at(1).simulate('click');
            postToVSCodeStub.should.have.been.calledOnceWithExactly({
                command: ExtensionCommands.SUBMIT_TRANSACTION,
                data: {
                    args: ['Green'],
                    channelName: 'mychannel',
                    evaluate: false,
                    namespace: 'GreenContract',
                    peerTargetNames: ['peer1'],
                    smartContract: 'greenContract',
                    transactionName: 'transactionOne',
                    transientData: '{"some": "data"}',
                    txDataFile: undefined,
                }
            });
        });

        it('should attempt to submit a transaction with custom peers when the evaluate button is clicked', () => {
            component = updateManualInputValues(component, 'transactionOne', '{"key": "Green"}', '{"some": "data"}');

            component = updateSharedInputValues(component, [{
                id: 'peer1',
                label: 'peer1'
            }]);

            component.find('#evaluate-button').at(1).simulate('click');
            postToVSCodeStub.should.have.been.calledOnceWithExactly({
                command: ExtensionCommands.EVALUATE_TRANSACTION,
                data: {
                    args: ['Green'],
                    channelName: 'mychannel',
                    evaluate: true,
                    namespace: 'GreenContract',
                    peerTargetNames: ['peer1'],
                    smartContract: 'greenContract',
                    transactionName: 'transactionOne',
                    transientData: '{"some": "data"}',
                    txDataFile: undefined,
                }
            });
        });

        it('should not submit if no transaction has been selected as the submit button is disabled', () => {
            component = updateManualInputValues(component, undefined, '{"key": "Green"}', '{"some": "data"}');
            component.find('#submit-button').at(1).simulate('click');
            postToVSCodeStub.should.not.have.been.called;
        });

        it('should not submit if no transaction has been selected as the evaluate button is disabled', () => {
            component = updateManualInputValues(component, undefined, '{"key": "Green"}', '{"some": "data"}');
            component.find('#evaluate-button').at(1).simulate('click');
            postToVSCodeStub.should.not.have.been.called;
        });
    });

    describe('Data file input', () => {
        const selectTxData: string = '[data-testid="select-tx-data-dir"]';

        describe('Without associatedTxdata', () => {
            beforeEach(() => {
                component = mount(<TransactionInputContainer smartContract={greenContract} associatedTxdata={undefined} txdataTransactions={txdataTransactions} preselectedTransaction={preselectedTransaction}/>);
                component = toggleContentSwitcher(component);
            });

            it('should show the Button and hide the FileUploaderItem', () => {
                expect(component.find(FileUploaderItem).exists()).toBeFalsy();
                expect(component.find(selectTxData).exists()).toBeTruthy();
            });

            it('should call the ASSOCIATE_TRANSACTION_DATA_DIRECTORY vscode command when the Button is clicked', () => {
                component.find(selectTxData).at(1).simulate('click');
                const { label, name, channel } = greenContract;
                postToVSCodeStub.should.have.been.calledOnceWithExactly({
                    command: ExtensionCommands.ASSOCIATE_TRANSACTION_DATA_DIRECTORY,
                    data: {
                        label,
                        name,
                        channel,
                    }
                });
            });

            it('should show the transaction dropdown as disabled and have no items as no directory is associated', () => {
                const dropdown: any = component.find(transactionNameSelector);
                expect(dropdown.prop('disabled')).toBeTruthy();
                expect(dropdown.prop('items')).toEqual([]);
            });

            it('should not submit if no directory is associated as the submit button is disabled', () => {
                component.find('#submit-button').at(1).simulate('click');
                postToVSCodeStub.should.not.have.been.called;
            });

            it('should not submit if no directory is associated as the evaluate button is disabled', () => {
                component.find('#evaluate-button').at(1).simulate('click');
                postToVSCodeStub.should.not.have.been.called;
            });
        });

        describe('With associatedTxdata', () => {
            beforeEach(() => {
                component = mount(<TransactionInputContainer smartContract={greenContract} associatedTxdata={associatedTxdata} txdataTransactions={txdataTransactions} preselectedTransaction={preselectedTransaction} />);
                component = toggleContentSwitcher(component);
            });

            it('should show the FileUploaderItem and hide the Button', () => {
                expect(component.find(FileUploaderItem).exists()).toBeTruthy();
                expect(component.find(selectTxData).exists()).toBeFalsy();
            });

            it('should call the DISSOCIATE_TRANSACTION_DATA_DIRECTORY vscode command when the onDelete prop is called on the FileUploaderItem', () => {
                act(() => {
                    component.find(FileUploaderItem).prop('onDelete')();
                });
                const { label, name, channel } = greenContract;
                postToVSCodeStub.should.have.been.calledOnceWithExactly({
                    command: ExtensionCommands.DISSOCIATE_TRANSACTION_DATA_DIRECTORY,
                    data: {
                        label,
                        name,
                        channel,
                    }
                });
            });

            it('should show the transaction dropdown as enabled and display the txdataTransactions as a directory is associated', () => {
                const dropdown: any = component.find(transactionNameSelector);
                expect(dropdown.prop('disabled')).toBeFalsy();
                expect(dropdown.prop('items')).toEqual(txdataTransactions);
            });

            it('updates the dropdown when the transaction is chosen', () => {
                let dropdown: any = component.find(transactionNameSelector);

                const defaultTransaction: IDataFileTransaction = {
                    transactionName: 'Select the transaction name',
                    transactionLabel: '',
                    arguments: [],
                    transientData: {},
                    txDataFile: '',
                };
                expect(dropdown.prop('selectedItem')).toEqual(defaultTransaction);

                component = updateDataInputValues(component, txdataTransactions[0]);

                dropdown = component.find(transactionNameSelector);
                expect(dropdown.prop('selectedItem')).toEqual(txdataTransactions[0]);
            });

            it('should attempt to submit a transaction when the submit button is clicked ', () => {
                component = updateDataInputValues(component, txdataTransactions[0]);

                component.find('#submit-button').at(0).simulate('click');

                postToVSCodeStub.should.have.been.calledOnceWithExactly({
                    command: ExtensionCommands.SUBMIT_TRANSACTION,
                    data: {
                        args: ['arg1', 'arg2'],
                        channelName: 'mychannel',
                        evaluate: false,
                        namespace: 'GreenContract',
                        peerTargetNames: greenContract.peerNames,
                        smartContract: 'greenContract',
                        transactionName: 'transactionOne',
                        transientData: { key: 'value' },
                        txDataFile: 'transactionData.txdata',
                    }
                });
            });

            it('should attempt to evaluate a transaction when the submit button is clicked ', () => {
                component = updateDataInputValues(component, txdataTransactions[0]);

                component.find('#evaluate-button').at(1).simulate('click');

                postToVSCodeStub.should.have.been.calledOnceWithExactly({
                    command: ExtensionCommands.EVALUATE_TRANSACTION,
                    data: {
                        args: ['arg1', 'arg2'],
                        channelName: 'mychannel',
                        evaluate: true,
                        namespace: 'GreenContract',
                        peerTargetNames: greenContract.peerNames,
                        smartContract: 'greenContract',
                        transactionName: 'transactionOne',
                        transientData: { key: 'value' },
                        txDataFile: 'transactionData.txdata',
                    }
                });
            });

            it('should attempt to submit a transaction with custom peers when the submit button is clicked', () => {
                component = updateDataInputValues(component, txdataTransactions[0]);

                component = updateSharedInputValues(component, [{
                    id: 'peer1',
                    label: 'peer1'
                }]);

                component.find('#submit-button').at(1).simulate('click');
                postToVSCodeStub.should.have.been.calledOnceWithExactly({
                    command: ExtensionCommands.SUBMIT_TRANSACTION,
                    data: {
                        args: ['arg1', 'arg2'],
                        channelName: 'mychannel',
                        evaluate: false,
                        namespace: 'GreenContract',
                        peerTargetNames: ['peer1'],
                        smartContract: 'greenContract',
                        transactionName: 'transactionOne',
                        transientData: { key: 'value' },
                        txDataFile: 'transactionData.txdata',
                    }
                });
            });

            it('should attempt to submit a transaction with custom peers when the evaluate button is clicked', () => {
                component = updateDataInputValues(component, txdataTransactions[0]);

                component = updateSharedInputValues(component, [{
                    id: 'peer1',
                    label: 'peer1'
                }]);

                component.find('#evaluate-button').at(1).simulate('click');
                postToVSCodeStub.should.have.been.calledOnceWithExactly({
                    command: ExtensionCommands.EVALUATE_TRANSACTION,
                    data: {
                        args: ['arg1', 'arg2'],
                        channelName: 'mychannel',
                        evaluate: true,
                        namespace: 'GreenContract',
                        peerTargetNames: ['peer1'],
                        smartContract: 'greenContract',
                        transactionName: 'transactionOne',
                        transientData: { key: 'value' },
                        txDataFile: 'transactionData.txdata',
                    }
                });
            });

            it('should not submit if no transaction has been selected as the submit button is disabled', () => {
                component.find('#submit-button').at(1).simulate('click');
                postToVSCodeStub.should.not.have.been.called;
            });

            it('should not submit if no transaction has been selected as the evaluate button is disabled', () => {
                component.find('#evaluate-button').at(1).simulate('click');
                postToVSCodeStub.should.not.have.been.called;
            });
        });
    });
});
