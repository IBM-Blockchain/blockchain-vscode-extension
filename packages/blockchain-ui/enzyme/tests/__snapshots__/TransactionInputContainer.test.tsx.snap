// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionInputContainer component should render the expected snapshot 1`] = `
Array [
  <div
    className="bx--content-switcher transaction-input-content-switch"
    onChange={[Function]}
    role="tablist"
    selectionMode="manual"
  >
    <button
      aria-selected={true}
      className="bx--content-switcher-btn bx--content-switcher--selected"
      data-testid="content-switch-manual"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="tab"
      tabIndex="0"
    >
      <span
        className="bx--content-switcher__label"
      >
        Manual input
      </span>
    </button>
    <button
      aria-selected={false}
      className="bx--content-switcher-btn"
      data-testid="content-switch-data"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="tab"
      tabIndex="-1"
    >
      <span
        className="bx--content-switcher__label"
      >
        Transaction data directory
      </span>
    </button>
  </div>,
  <form
    className="bx--form"
    id="create-txn-form"
  >
     
    <div
      className="bx--dropdown__wrapper bx--list-box__wrapper"
    >
      <span
        className="bx--label"
      >
        Transaction name
      </span>
      <div
        aria-label="dropdown"
        className="bx--dropdown bx--list-box"
        id="transaction-select"
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        <div
          aria-controls={null}
          aria-disabled={false}
          aria-expanded={false}
          aria-haspopup={true}
          aria-labelledby="dropdown-label-transaction-select dropdown-field-label-transaction-select"
          aria-owns={null}
          className="bx--list-box__field"
          data-toggle={true}
          onBlur={[Function]}
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          tabIndex="0"
          type="button"
        >
          <span
            className="bx--list-box__label"
            htmlFor="downshift-0-input"
            id="dropdown-field-label-transaction-select"
          >
            Select the transaction name
          </span>
          <div
            className="bx--list-box__menu-icon"
          >
            <svg
              aria-label="Open menu"
              focusable="false"
              height={16}
              name="chevron--down"
              preserveAspectRatio="xMidYMid meet"
              role="img"
              viewBox="0 0 16 16"
              width={16}
              xmlns="http://www.w3.org/2000/svg"
            >
              <polygon
                points="8,11 3,6 3.7,5.3 8,9.6 12.3,5.3 13,6"
              />
              <title>
                Open menu
              </title>
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      className="bx--form-item"
    >
      <label
        className="bx--label bx--label--disabled"
        htmlFor="arguments-text-area"
      >
        Transaction arguments
      </label>
      <div
        className="bx--text-area__wrapper"
        data-invalid={null}
      >
        <textarea
          aria-describedby={null}
          aria-invalid={null}
          className="bx--text-area"
          cols={50}
          disabled={true}
          id="arguments-text-area"
          onChange={[Function]}
          onClick={[Function]}
          placeholder=""
          rows={4}
          value=""
        />
      </div>
    </div>
    <div
      className="bx--form-item"
    >
      <label
        className="bx--label"
        htmlFor="transient-data-input"
      >
        Transient data (optional)
      </label>
      <div
        className="bx--text-area__wrapper"
        data-invalid={null}
      >
        <textarea
          aria-describedby={null}
          aria-invalid={null}
          className="bx--text-area"
          cols={50}
          disabled={false}
          id="transient-data-input"
          onChange={[Function]}
          onClick={[Function]}
          placeholder=""
          rows={4}
          value=""
        />
      </div>
    </div>
    <div
      className="bx--multi-select__wrapper bx--list-box__wrapper"
    >
      <span
        className="bx--label"
        id="multiselect-label-1"
      >
        Target specific peer (optional)
      </span>
      <div
        className="bx--multi-select bx--multi-select--selected bx--list-box"
        id="peer-select"
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        <div
          aria-controls={null}
          aria-disabled={false}
          aria-expanded={false}
          aria-haspopup={true}
          aria-labelledby="multiselect-label-1 multiselect-field-label-1"
          aria-owns={null}
          className="bx--list-box__field"
          data-toggle={true}
          onBlur={[Function]}
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          tabIndex="0"
          type="button"
        >
          <div
            aria-label="Clear Selection"
            className="bx--list-box__selection bx--tag--filter bx--list-box__selection--multi"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={0}
            title="Clear all selected items"
          >
            2
            <svg
              aria-hidden={true}
              focusable="false"
              height={16}
              preserveAspectRatio="xMidYMid meet"
              viewBox="0 0 32 32"
              width={16}
              xmlns="http://www.w3.org/2000/svg"
            >
              <polygon
                points="24 9.4 22.6 8 16 14.6 9.4 8 8 9.4 14.6 16 8 22.6 9.4 24 16 17.4 22.6 24 24 22.6 17.4 16 24 9.4"
              />
            </svg>
          </div>
          <span
            className="bx--list-box__label"
            id="multiselect-field-label-1"
          >
            Select peers
          </span>
          <div
            className="bx--list-box__menu-icon"
          >
            <svg
              aria-label="Open menu"
              focusable="false"
              height={16}
              name="chevron--down"
              preserveAspectRatio="xMidYMid meet"
              role="img"
              viewBox="0 0 16 16"
              width={16}
              xmlns="http://www.w3.org/2000/svg"
            >
              <polygon
                points="8,11 3,6 3.7,5.3 8,9.6 12.3,5.3 13,6"
              />
              <title>
                Open menu
              </title>
            </svg>
          </div>
        </div>
      </div>
    </div>
    <fieldset
      className="bx--fieldset"
      id="submit-and-evaluate-buttons"
    >
      <legend
        className="bx--label"
      >
        
      </legend>
      <div
        className="submit-txn-button-container"
      >
        <button
          className="submit-txn-button bx--btn bx--btn--secondary bx--btn--disabled"
          disabled={true}
          id="evaluate-button"
          onClick={[Function]}
          tabIndex={0}
          type="button"
        >
          Evaluate transaction
        </button>
        <button
          className="submit-txn-button bx--btn bx--btn--primary bx--btn--disabled"
          disabled={true}
          id="submit-button"
          onClick={[Function]}
          tabIndex={0}
          type="button"
        >
          Submit transaction
        </button>
      </div>
    </fieldset>
     
  </form>,
]
`;
